clippy_lints/src/modulo_arithmetic.rs

```diff
@@ -0,0 +1,106 @@
+    /// depending on the language, when negative numbers are involved.
+    /// If you interop with different languages it might be beneficial
+    /// to double check all places that use modulo arithmetic.
+    ///
+    /// For example, in Rust `17 % -3 = 2`, but in Python `17 % -3 = -1`.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    /// ```rust
+    /// let x = -17 % 3;
+    /// ```
+    pub MODULO_ARITHMETIC,
+    restriction,
+    "any modulo arithmetic statement"
+}
+
+declare_lint_pass!(ModuloArithmetic => [MODULO_ARITHMETIC]);
+
+fn is_negative(operand: &Expr, cx: &LateContext<'_, '_>, expr: &Expr) -> Option<(bool, Option<String>)> {
+    match constant(cx, cx.tables, operand) {
+        Some((Constant::Int(v), _)) => match cx.tables.expr_ty(expr).kind {
+            ty::Int(ity) => {
+                let value = sext(cx.tcx, v, ity);
+                return Some((value < 0, Some(value.to_string())));
+            },
+            ty::Uint(_) => {
+                return Some((false, None));
+            },
+            _ => {},
+        },
+        Some((Constant::F32(f), _)) => {
+            return Some((f < 0.0, Some(format!("{:.3}", f))));
+        },
+        Some((Constant::F64(f), _)) => {
+            return Some((f < 0.0, Some(format!("{:.3}", f))));
+        },
+        _ => {},
+    }
+    None
+}
+
+fn might_have_negative_value(t: &ty::TyS<'_>) -> bool {
+    t.is_signed() || t.is_floating_point()
+}
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for ModuloArithmetic {
+    fn check_expr(&mut self, cx: &LateContext<'a, 'tcx>, expr: &'tcx Expr) {
+        match &expr.kind {
+            ExprKind::Binary(op, lhs, rhs) | ExprKind::AssignOp(op, lhs, rhs) => {
+                if let BinOpKind::Rem = op.node {
+                    let lhs_negative = is_negative(lhs, cx, expr);
+                    let rhs_negative = is_negative(rhs, cx, expr);
+                    if lhs_negative.is_some() && rhs_negative.is_some() {
+                        // Operands are consts
+                        if_chain! {
+                            if let Some(lhs_negative) = lhs_negative;
+                            if let Some(rhs_negative) = rhs_negative;
+                            if lhs_negative.0 ^ rhs_negative.0;
+                            then
+                            {
+                                span_note_and_lint(
+                                    cx,
+                                    MODULO_ARITHMETIC,
+                                    expr.span,
+                                    &format!(
+                                        "you are using modulo operator on constants with different signs: `{} % {}`", lhs_negative.1.unwrap(), rhs_negative.1.unwrap()
+                                    ),
+                                    expr.span,
+                                    "double check for expected result especially when interoperating with different languages",
```

Comment:
What are your thoughts on also suggesting usage of `rem_euclid` or `checked_rem_euclid` when the signs are different? Sometime that's what the user intended when performing modulo arithmetic on signed integers and they may not be aware that this method exists.

Reply:
Well, I myself was not aware that such functions exist :)
Let me check how they relate to the content of this PR.

Reply:
OK, so I see that `wrapping_rem_euclid()` and `checked_rem_euclid()` yield results that are consistent with what we can see in Python.

I will add info about these methods in the lint note later today.

Thanks!

Reply:
Yup these methods are only for integer types so we would only want to suggest these when operating on integers.

Reply:
Relevant changes are now pushed to the branch.